Replace `rest` field in ParamDef with special members

gzm0 opened the pull request.

gzm0 commented:
The bulk of the `ir` infrastructure isn't changed yet. I wanted to see how it affects the compiler/linker.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala:
```diff
@@ -458,40 +469,45 @@ private[emitter] class FunctionEmitter(sjsGen: SJSGen) {
-      val newParams =
-        (if (translateRestParam) params.init else params).map(transformParamDef)
+          case Some(restParam) =>
+            val newParams = normalParams :+ transformParamDef(restParam, rest = true)
```
If `!esFeatures.useECMAScript2015`, `newParams` should be only `normalParams`.

sjrd reviewed (commented):
The changes in the compiler seem a bit awkward, but the changes in the linker look like they are an improvement, so that seems overall reasonable to me.

sjrd commented on ir/shared/src/main/scala/org/scalajs/ir/Serializers.scala:
```diff
@@ -770,18 +770,24 @@ object Serializers {
-      buffer.writeInt(paramDefs.size)
-      paramDefs.foreach(writeParamDef)
+    def writeParamDefs(paramDefs: List[ParamDef],
+        restParamDef: Option[ParamDef] = None): Unit = {
```
Consider adding a `// temp` comment telling that this tests backward compat. I was quite confused for a while that this method implemented the old scheme whereas the reading would be trying to read the new scheme.

Same in `writeParamDef`, I guess.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala:
```diff
@@ -511,15 +512,30 @@ private[emitter] class FunctionEmitter(sjsGen: SJSGen) {
+          js.Function(actualArrowFun, normalParams, cleanedNewBody)
+
+        case Some(restParam) if esFeatures.useECMAScript2015 =>
+          val params =
+            normalParams :+ transformParamDef(restParam, rest = true)
+
+          js.Function(actualArrowFun, params, cleanedNewBody)
```
Have you considered also applying the same refactoring to the `javascript.Tree`s? It seems like it would be more consistent and cleaner.

gzm0 replied:
Yep. I'll do that (likely in a separate commit).

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala:
```diff
@@ -881,17 +885,17 @@ private[optimizer] abstract class OptimizerCore(config: CommonPhaseConfig) {
           }
 
-          if (!arrow || params.exists(_.rest)) {
+          if (!arrow || restParam.nonEmpty) {
```
`restParam.isDefined` is perhaps bit more canonical, and even clearer in this case?

gzm0 commented on ir/shared/src/main/scala/org/scalajs/ir/Serializers.scala:
```diff
@@ -770,18 +770,24 @@ object Serializers {
     }
 
-    def writeParamDef(paramDef: ParamDef): Unit = {
+    def writeParamDef(paramDef: ParamDef, rest: Boolean = false): Unit = {
```
I have, but because it was only temporary, this seemed better. LMK what you'd like me to do.

sjrd replied:
Right, that makes sense. I had not noticed that.

gzm0 commented:
I believe I have addressed all the comments. The js tree refactoring introduces quite a bunch of `None` in the CoreJSLib. LMK if you'd like me to build some DSL help there. (I'm unsure whether it is worth it or not).

sjrd commented:
> The js tree refactoring introduces quite a bunch of `None` in the CoreJSLib. LMK if you'd like me to build some DSL help there. (I'm unsure whether it is worth it or not).

Overloads of `globalFunctionDef`, `genArrowFunction`, and perhaps `js.Function.apply` and `js.MethodDef.apply`, that don't take a `restParam` argument would probably go a long way to getting rid of all those `None`s.

sjrd reviewed (changes requested):
Overall LGTM. I think we can avoid a lot of `None`s with a few overloads (see above).

gzm0 commented:
I have been thinking about introducing overloads as well (I even started with a `genArrowFunction` overload that didn't take the `restParam`). However, it seemed that they really were not useful outside of the CoreJSLib (feels like almost an equal amount pass `restParam` forward). So I'm a bit reluctant to introduce the overloads globally.

In the CoreJSLib it seems that most `None`s are in `genArrowFunction` and `MethodDef` (`globalFunctionDef` is only used directly if it is not in the core scope, so there are not that many usages).

I'm wondering if we should generalize the `defineFunctionN` approach (smth like `genLambdaN` and `genMethodN`). Alternatively, I could also introduce a local overload of `genArrowFunction` (and maybe `genMethodDef`) in CoreJSLib. WDYT?

sjrd commented:
> I'm wondering if we should generalize the `defineFunctionN` approach (smth like `genLambdaN` and `genMethodN`).

If it helps, then why not. Otherwise I think it's also fine if we keep it as is.

> Alternatively, I could also introduce a local overload of `genArrowFunction` (and maybe `genMethodDef`) in CoreJSLib.

I don't think that will work, because the local definition would shadow the other one, not overload it.

gzm0 commented:
I have gone with a simple definition of `genArrowFunction` that shadows the one in `JSGen`. In fact, we must not use rest parameters in the CoreJSLib (unconditionally) as long as we target ES 5.1. WDYT?

sjrd reviewed (approved):
LGTM!
